src/lein_monolith/task/fingerprint.clj

```diff
@@ -0,0 +1,352 @@
+  {::up-to-date ["is up-to-date" "are up-to-date" :green]
+   ::new-project ["is a new project" "are new projects" :red]
+   ::sources ["has updated sources" "have updated sources" :red]
+   ::deps ["has updated external dependencies" "have updated external dependencies" :yellow]
+   ::upstream ["is downstream of an affected project" "are downstream of affected projects" :yellow]
+   ::unknown ["has a different fingerprint" "have different fingerprints" :red]})
+
+
+(defn explain-str
+  [ctx marker project-name]
+  (let [[singular plural color] (reason-details (explain-kw ctx marker project-name))]
+    (ansi/sgr singular color)))
+
+
+(defn save!
+  "Save the fingerprints for a project with the specified marker."
+  [ctx marker project-name]
+  (let [current (fingerprints ctx project-name)]
+    (update-fingerprints-file!
+      (:monolith ctx)
+      assoc-in [marker project-name] current)))
+
+
+(defn- list-projects
+  [project-names color]
+  (->> project-names
+       (map #(ansi/sgr % color))
+       (str/join ", ")))
+
+
+(defn changed
+  [project opts markers]
+  (let [[monolith subprojects] (u/load-monolith! project)
+        ctx (context monolith subprojects)
+        targets (filter subprojects (target/select monolith subprojects opts))
+        markers (if (seq markers)
+                  markers
+                  (keys (:initial ctx)))]
+    (cond
+      (empty? markers) (lein/info "No saved fingerprint markers")
+      (empty? targets) (lein/info "No projects selected")
+
+      :else
+      (doseq [marker markers
+              :let [changed (->> targets
+                                 (filter (partial changed? ctx marker))
+                                 (set))
+                    pct-changed (if (seq targets)
+                                  (* 100.0 (/ (count changed) (count targets)))
+                                  0.0)]]
+        (lein/info (ansi/sgr (format "%.2f%%" pct-changed)
+                             (cond
+                               (== 0.0 pct-changed) :green
+                               (< pct-changed 50) :yellow
+                               :else :red))
+                   "out of"
+                   (count targets)
+                   "projects have out-of-date"
+                   (ansi/sgr marker :bold)
+                   "fingerprints:\n")
+        (let [reasons (group-by (partial explain-kw ctx marker) targets)]
+          (doseq [k [::unknown ::new-project ::sources ::resources ::deps ::upstream ::up-to-date]]
+            (when-let [projs (seq (k reasons))]
+              (let [[singular plural color] (reason-details k)
+                    c (count projs)]
+                (lein/info "*" (ansi/sgr (count projs) color)
+                           (str (if (= 1 c) singular plural)
+                                (when-not (#{::up-to-date ::upstream} k)
+                                  (str ": " (list-projects projs color)))))))))
+        (when (seq (rest markers))
```

Comment:
This isn't a loop, so `markers` isn't going to change here.

Reply:
I realized that while writing this and figured an extra newline at the very end is fine in this case, but I guess I should instead figure out a way to write this more elegantly.